File: 100milliongold/leetcode/tree/104.Maximum Depth of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 var maxDepth = function(root) {
    return searchDepth(root , 0)
};

const searchDepth = (node, depth) => {
    if(node === null) return depth;
    depth++;
    
    const leftDepth = searchDepth(node.left, depth);
    const rightDepth = searchDepth(node.right, depth);
    
    const max = Math.max(leftDepth, rightDepth);
    return max;
}

File: gunkims/leetcode/tree/104.Maximum Depth of Binary Tree.java
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.maxDepth(new TreeNode(1, new TreeNode(1, new TreeNode(3, new TreeNode(4, null, null), null), null), new TreeNode(3))));
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    public int solve(TreeNode node, int depth) {
        if(node == null) {
            return depth;
        }
        return Math.max(solve(node.left, depth+1), solve(node.right, depth+1));
    }
    public int maxDepth(TreeNode root) {
        int res = solve(root, 0);
        return res;
    }
}